File: src/test/java/nl/wernerkroneman/Drawy/ConcreteModelling/AbstractToConcreteTest.java
package nl.wernerkroneman.Drawy.ConcreteModelling;

import nl.wernerkroneman.Drawy.AbstractToConcreteConverter.AbstractToConcrete;
import nl.wernerkroneman.Drawy.AbstractToConcreteConverter.PositionalSolver;
import nl.wernerkroneman.Drawy.Modelling.*;
import org.joml.Vector3d;
import org.junit.Assert;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static nl.wernerkroneman.Drawy.Modelling.RelativePositionConstraint.*;

/**
 * Created by werner on 27-12-16.
 */
public class AbstractToConcreteTest {
    @Test
    public void getTranslationToFit() throws Exception {

        MeshFactory meshFactory = new DefaultMeshFactory();
        PrimitiveGenerator primitiveGenerator = new PrimitiveGenerator(meshFactory);

        SceneNode root = new SceneNode();

        SceneNode child = new SceneNode();
        root.addChild(child);

        child.addDrawable(new Drawable(primitiveGenerator.generateUnitCube()));

        AbstractToConcrete converter = new AbstractToConcrete(meshFactory);

        Vector3d trans = converter.getTranslationToFit(child, child.computeLocalAABB());
        Assert.assertEquals(0, trans.length(), 0.01);

        Vector3d trans2 = converter.getTranslationToFit(child, new AABB(new Vector3d(-50), new Vector3d(-51)));
        Assert.assertEquals(-50.5, trans2.x, 0.01);
        Assert.assertEquals(-50.5, trans2.y, 0.01);
        Assert.assertEquals(-50.5, trans2.z, 0.01);

    }

    @Test
    public void findEmptyPlaceTest() throws Exception {

        MeshFactory meshFactory = new DefaultMeshFactory();
        PrimitiveGenerator primitiveGenerator = new PrimitiveGenerator(meshFactory);

        for (int i = 0;
             i < 100;
             i++) {
            SceneNode root = new SceneNode();

            SceneNode child = new SceneNode();
            root.addChild(child);

            child.addDrawable(new Drawable(primitiveGenerator.generateUnitCube()));

            AbstractToConcrete converter = new AbstractToConcrete(meshFactory);

            List<AABB> doesNotIntersect = new ArrayList<>();
            doesNotIntersect.add(child.computeWorldAABB());

            AABB empty = PositionalSolver.findEmptyPlace(doesNotIntersect, 0, new Vector3d(2),
                    new AABB(new Vector3d(Double.POSITIVE_INFINITY),
                             new Vector3d(Double.NEGATIVE_INFINITY)));

            Assert.assertFalse(empty.intersects(new AABB(new Vector3d(0.5), new Vector3d(-0.5)), 0));

            Assert.assertEquals(2, empty.getSizeZ(), 0.01);
            Assert.assertEquals(2, empty.getSizeX(), 0.01);
            Assert.assertEquals(2, empty.getSizeY(), 0.01);

            Assert.assertTrue(empty.getCenter(new Vector3d()).length() < 10);

        }
    }

    @Test
    public void computeSceneSingleCube() throws Exception {
        // This should produce a single cube centered at (0,0,0) of edge length 1.

        PrimitiveModel cube = new PrimitiveModel(PrimitiveModel.ShapeType.CUBE, "Cube");

        AbstractToConcrete converter = new AbstractToConcrete(new DefaultMeshFactory());

        Scene result = converter.computeScene(cube);

        Assert.assertEquals(1, result.getRootSceneNode().getDrawables().size());

        Mesh cubemesh = result.getRootSceneNode().getDrawables().get(0).getMesh();

        AABB box = cubemesh.computeAABB();

        Vector3d center = box.getCenter(new Vector3d());
        Assert.assertEquals(0, center.x, 0.01);
        Assert.assertEquals(0, center.y, 0.01);
        Assert.assertEquals(0, center.z, 0.01);

        Assert.assertEquals(1, box.getSizeY(), 0.01);
        Assert.assertEquals(1, box.getSizeZ(), 0.01);
        Assert.assertEquals(1, box.getSizeX(), 0.01);
    }

    @Test
    public void computeSceneSingleCubeInComposite() throws Exception {
        // This should produce a single cube centered at (0,0,0) of edge length 1.

        PrimitiveModel cube = new PrimitiveModel(PrimitiveModel.ShapeType.CUBE, "Cube");

        CompositeModel composite = new CompositeModel("Cube container.");

        composite.addComponentForModel(cube);

        AbstractToConcrete converter = new AbstractToConcrete(new DefaultMeshFactory());

        Scene result = converter.computeScene(composite);

        Assert.assertEquals(0, result.getRootSceneNode().getDrawables().size());
        Assert.assertEquals(1, result.getRootSceneNode().getChildren().size());

        SceneNode childNode = result.getRootSceneNode().getChildren().get(0);
        Assert.assertEquals(1, childNode.getDrawables().size());

        Mesh cubemesh = childNode.getDrawables().get(0).getMesh();

        AABB box = cubemesh.computeAABB();

        Vector3d center = box.getCenter(new Vector3d());
        Assert.assertEquals(0, center.x, 0.01);
        Assert.assertEquals(0, center.y, 0.01);
        Assert.assertEquals(0, center.z, 0.01);

        Assert.assertEquals(1, box.getSizeY(), 0.01);
        Assert.assertEquals(1, box.getSizeZ(), 0.01);
        Assert.assertEquals(1, box.getSizeX(), 0.01);
    }

    @Test
    public void computeSceneMultipleCubeInComposite() throws Exception {
        // This should produce multiple, non-overlapping cubes.
        for (int rep = 0; rep < 50; rep++) {
            PrimitiveModel cube = new PrimitiveModel(PrimitiveModel.ShapeType.CUBE, "Cube");

            CompositeModel composite = new CompositeModel("Cube container.");

            final int NUM_CUBES = 10;

            for (int i = 0; i < NUM_CUBES; i++) {
                composite.addComponentForModel(cube);
            }

            AbstractToConcrete converter = new AbstractToConcrete(new DefaultMeshFactory());

            Scene result = converter.computeScene(composite);

            Assert.assertEquals(0, result.getRootSceneNode().getDrawables().size());
            Assert.assertEquals(NUM_CUBES, result.getRootSceneNode().getChildren().size());

            for (int i = 0;
                 i < NUM_CUBES;
                 i++) {
                for (int j = 0;
                     j < NUM_CUBES;
                     j++) {
                    if (i != j) {

                        AABB a = result.getRootSceneNode().getChildren().get(i).getDrawables().get(0).getWorldAABB();
                        AABB b = result.getRootSceneNode().getChildren().get(j).getDrawables().get(0).getWorldAABB();

                        Assert.assertFalse(a.intersects(b, 0));
                    }
                }
            }
        }
    }

    /**
     * Supposed to create a stack of 3 cubes.
     */
    @Test
    public void testRelativePositionConstraint() {
        for (int rep = 0; rep < 100; rep++) {
            // Set up a simple scene with two cubes
            PrimitiveModel cube = new PrimitiveModel(PrimitiveModel.ShapeType.CUBE, "Cube");

            CompositeModel composite = new CompositeModel("Cube container.");

            // Create two cube components.
            CompositeModel.Component cubeA = composite.addComponentForModel(cube);
            CompositeModel.Component cubeB = composite.addComponentForModel(cube);
            CompositeModel.Component cubeC = composite.addComponentForModel(cube);

            composite.getConstraints().add(new RelativePositionConstraint(cubeA, cubeB, ABOVE, Distance.ANY));
            composite.getConstraints().add(new RelativePositionConstraint(cubeB, cubeC, ABOVE, Distance.ANY));

            // Compute the concrete scene
            AbstractToConcrete converter = new AbstractToConcrete(new DefaultMeshFactory());

            Scene scene = converter.computeScene(composite);

            // Should be three children
            List<SceneNode> children = scene.getRootSceneNode().getChildren();
            Assert.assertEquals(3, children.size());

            // Check for intersecting pairs
            for (int i = 0; i < 3; i++) {
                for (int j = 0; j < 3; j++) {
                    if (i != j) {
                        AABB boxA = children.get(i).computeWorldAABB();
                        AABB boxB = children.get(j).computeWorldAABB();
                        Assert.assertFalse(boxA.intersects(boxB, 0.01));
                    }
                }
            }

            // Check whether the bounding box is consistent with a stack of 3 cubes.
            AABB totalBox = scene.getRootSceneNode().computeWorldAABB();
            Assert.assertEquals(3, totalBox.getSizeY(), 0.5);
            Assert.assertEquals(1, totalBox.getSizeX(), 0.1);
            Assert.assertEquals(1, totalBox.getSizeZ(), 0.1);
        }
    }

    /**
     * Supposed to create a stack of 3 cubes.
     */
    @Test
    public void testRelativePositionConstraintExactDistance() {
        for (int rep = 0; rep < 50; rep++) {
            // Set up a simple scene with two cubes
            PrimitiveModel cube = new PrimitiveModel(PrimitiveModel.ShapeType.CUBE, "Cube");

            CompositeModel composite = new CompositeModel("Cube container.");

            // Create two cube components.
            CompositeModel.Component cubeA = composite.addComponentForModel(cube);
            CompositeModel.Component cubeB = composite.addComponentForModel(cube);
            CompositeModel.Component cubeC = composite.addComponentForModel(cube);

            composite.getConstraints().add(new RelativePositionConstraint(cubeA, cubeB, ABOVE, new FixedDistance(1)));
            composite.getConstraints().add(new RelativePositionConstraint(cubeB, cubeC, ABOVE, new FixedDistance(2)));

            // Compute the concrete scene
            AbstractToConcrete converter = new AbstractToConcrete(new DefaultMeshFactory());

            Scene scene = converter.computeScene(composite);

            // Should be three children
            List<SceneNode> children = scene.getRootSceneNode().getChildren();
            Assert.assertEquals(3, children.size());

            // Check for intersecting pairs
            for (int i = 0; i < 3; i++) {
                for (int j = 0; j < 3; j++) {
                    if (i != j) {
                        AABB boxA = children.get(i).computeWorldAABB();
                        AABB boxB = children.get(j).computeWorldAABB();
                        Assert.assertFalse(boxA.intersects(boxB, 0.01));
                    }
                }
            }

            // Check whether the bounding box is consistent with a stack of 3 cubes.
            AABB totalBox = scene.getRootSceneNode().computeWorldAABB();
            Assert.assertEquals(6, totalBox.getSizeY(), 0.5);
            Assert.assertEquals(1, totalBox.getSizeX(), 0.1);
            Assert.assertEquals(1, totalBox.getSizeZ(), 0.1);
        }
    }

}

File: src/test/java/nl/wernerkroneman/Drawy/ConcreteModelling/AbstractToConcreteTest.kt
/*
 * Copyright (c) 2017 Werner Kroneman
 *
 * This file is part of DrawyMcDrawface.
 *
 * DrawyMcDrawface is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DrawyMcDrawface is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DrawyMcDrawface.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.wernerkroneman.Drawy.ConcreteModelling

import nl.wernerkroneman.Drawy.AbstractToConcreteConverter.AbstractToConcrete
import nl.wernerkroneman.Drawy.Modelling.*
import org.joml.Vector3d
import org.junit.Assert
import org.junit.Test

class AbstractToConcreteTest {

    // This should produce a single cube centered at (0,0,0) of edge length 1.
    val cube = ModelSpecification(shape = Shape.CUBE)

    @Test
    @Throws(Exception::class)
    fun computeSceneSingleCube() {

        val converter = AbstractToConcrete(DefaultMeshFactory())

        val box = converter.concreteForModel(cube).aabb

        val center = box.center
        Assert.assertEquals(0.0, center.x, 0.01)
        Assert.assertEquals(0.0, center.y, 0.01)
        Assert.assertEquals(0.0, center.z, 0.01)

        Assert.assertEquals(1.0, box.sizeX, 0.01)
        Assert.assertEquals(1.0, box.sizeY, 0.01)
        Assert.assertEquals(1.0, box.sizeZ, 0.01)
    }

    @Test
    @Throws(Exception::class)
    fun computeSceneSingleCubeInComposite() {
        // This should produce a single cube centered at (0,0,0) of edge lateral 1.
        val composite = ModelSpecification(has = listOf(cube))

        val result = AbstractToConcrete(DefaultMeshFactory()).concreteForModel(composite)

        Assert.assertEquals(1, result.children.size)
        Assert.assertEquals(1, result.children.first().drawables.size)

        val box = result.aabb

        val center = box.center
        Assert.assertEquals(0.0, center.x, 0.01)
        Assert.assertEquals(0.0, center.y, 0.01)
        Assert.assertEquals(0.0, center.z, 0.01)

        Assert.assertEquals(1.0, box.sizeY, 0.01)
        Assert.assertEquals(1.0, box.sizeZ, 0.01)
        Assert.assertEquals(1.0, box.sizeX, 0.01)
    }

    /*@Test
    @Throws(Exception::class)
    fun computeSceneMultipleCubeInComposite() {
        // This should produce multiple, non-overlapping cubes.
        for (rep in 0..49) {
            val cube = PrimitiveModelSpecificationBase(name = "Cube", shape = PrimitiveModelSpecification.ShapeType.CUBE)

            val composite = CompositeModelSpecificationBase(name = "Cube container.")

            val NUM_CUBES = 10

            for (i in 0..NUM_CUBES - 1) {
                composite.directComponents.add(cube.derive("Cube ${i + 1}").apply {
                    this.location = NOT_INTERSECTING_SOLID
                })
            }

            val converter = AbstractToConcrete(DefaultMeshFactory())

            val result = converter.computeScene(composite)

            Assert.assertEquals(0, result.rootSceneNode.getDrawables().size.toLong())
            Assert.assertEquals(NUM_CUBES.toLong(), result.rootSceneNode.getChildren().size.toLong())

            (0..NUM_CUBES - 1).forEach { i ->
                (0..NUM_CUBES - 1).forEach { j ->
                    if (i != j) {
                        val a = result.rootSceneNode.getChildren()[i].getDrawables()[0].worldAABB
                        val b = result.rootSceneNode.getChildren()[j].getDrawables()[0].worldAABB

                        Assert.assertFalse(a.intersects(b, 0.01))
                    }
                }
            }
        }
    }

    /**
     * Supposed to create a stack of 3 cubes.
     */
    @Test
    fun testRelativePositionConstraint() {
        for (rep in 0..99) {
            // Set up a simple scene with two cubes
            val cube = PrimitiveModelSpecificationBase(name = "Cube", shape = PrimitiveModelSpecification.ShapeType.CUBE)

            val composite = CompositeModelSpecificationBase(name = "Cube container.")

            // Create three cube directComponents.
            val cubeA = cube.derive("A")
            composite.directComponents.add(cubeA)

            val cubeB = cube.derive("B")
            composite.directComponents.add(cubeB)

            val cubeC = cube.derive("C")
            composite.directComponents.add(cubeC)

            cubeA.location = RelativeLocation(cubeB, ABOVE, Distance.ANY)
            cubeB.location = RelativeLocation(cubeC, ABOVE, Distance.ANY)

            // Compute the concrete scene
            val converter = AbstractToConcrete(DefaultMeshFactory())

            val scene = converter.computeScene(composite)

            // Should be three children
            val children = scene.rootSceneNode.getChildren()
            Assert.assertEquals(3, children.size.toLong())

            // Check for intersecting pairs
            for (i in 0..2) {
                for (j in 0..2) {
                    if (i != j) {
                        val boxA = children[i].computeWorldAABB()
                        val boxB = children[j].computeWorldAABB()
                        Assert.assertFalse(boxA.intersects(boxB, 0.01))
                    }
                }
            }

            // Check whether the bounding box is consistent with a stack of 3 cubes.
            val totalBox = scene.rootSceneNode.computeWorldAABB()
            Assert.assertEquals(3.0, totalBox.sizeY, 0.5)
            Assert.assertEquals(1.0, totalBox.sizeX, 0.1)
            Assert.assertEquals(1.0, totalBox.sizeZ, 0.1)
        }
    }

    /**
     * Supposed to create a stack of 3 cubes.
     */
    @Test
    fun testRelativePositionConstraintExactDistance() {
        for (rep in 0..49) {
            // Set up a simple scene with two cubes
            val cube = PrimitiveModelSpecificationBase(name = "Cube", shape = PrimitiveModelSpecification.ShapeType.CUBE)

            val composite = CompositeModelSpecificationBase(name = "Cube container.")

            // Create three cube directComponents.
            val cubeA = cube.derive("A")
            composite.directComponents.add(cubeA)

            val cubeB = cube.derive("B")
            composite.directComponents.add(cubeB)

            val cubeC = cube.derive("C")
            composite.directComponents.add(cubeC)

            cubeA.location = RelativeLocation(cubeB, ABOVE, FixedDistance(0.0))
            cubeB.location = RelativeLocation(cubeC, ABOVE, FixedDistance(0.0))

            // Compute the concrete scene
            val converter = AbstractToConcrete(DefaultMeshFactory())

            val scene = converter.computeScene(composite)

            // Should be three children
            val children = scene.rootSceneNode.getChildren()
            Assert.assertEquals(3, children.size.toLong())

            // Check for intersecting pairs
            for (i in 0..2) {
                for (j in 0..2) {
                    if (i != j) {
                        val boxA = children[i].computeWorldAABB()
                        val boxB = children[j].computeWorldAABB()
                        Assert.assertFalse(boxA.intersects(boxB, 0.01))
                    }
                }
            }

            // Check whether the bounding box is consistent with a stack of 3 cubes.
            val totalBox = scene.rootSceneNode.computeWorldAABB()
            Assert.assertEquals(3.0, totalBox.sizeY, 0.5)
            Assert.assertEquals(1.0, totalBox.sizeX, 0.1)
            Assert.assertEquals(1.0, totalBox.sizeZ, 0.1)
        }
    }*/

}